readthedocs/search/parsers.py

```diff
@@ -83,13 +83,7 @@ def _get_main_node(self, html):
         if main_node:
             return main_node
 
-        # TODO: this could be done in smarter way,
-        # checking for common parents between all h nodes.
-        first_header = body.css_first('h1')
-        if first_header:
-            return first_header.parent
```

Comment:
It seems we are changing this logic on this PR as well. It may be good to make this change in a separate set of changes than mixing them with the implementation of a new generic parser.